### DISCUSSION TITLE
How to get Initiator Chain?

### DISCUSSION THREAD

**Discussion by thejohnbridge** [cefsharp/CefSharp#4483] [Q&A]
Hello,
I am trying to retrieve the request initiator chain with CEF, the same one you see in the browser's developer tools - when you go to Network and then click on Initiator tab, but I cannot find this information.
More specifically:
- first I tried overriding **IRequestHandler.OnResourceLoadComplete**, my callback does get called at each URL loaded by the page and I do have access to IBrowser and IRequest, I can find the request's referrer, but using the referrer to rebuild the tree of URL's is not giving me the same results I see in the browser's Initiator Chain.
- I then used the DeveloperTools.Network.EnableAsync, I installed my handler for **Cefsharp.Devtools.Network.NetworkClient.RequestWillBeSent** callback, it gets invoked for every GET request and I do get the IRequest and the Initiator objects, but again they do not contain the information I need to build the Initiator chain. In many cases the Initiator URL is empty, in other cases it's just the main domain of the website.

As an example, if I browse slashdot.org I see this Initiator Chain in the browser (Chrome or Edge):

http://www.slashdot.org
..+-> https://www.slashdot.org
.............+-> https://slashdot.org
.........................+-> https://a.fsdn.com.sd.classic.css?e97300f6d7a821a0
....................................+- ...
.............+- ...
.............+- https://d3tglifpd8whs6.cloudfront.net/js/prebid/slash-homepage/slash-homepage.min.js
.....................+- ...
.....................+- https://securepubads.g.doubleclick.net/tag/js/gpt.js

Here, the browser tells me that https://securepubads.g.doubleclick.net/tag/js/gpt.js is initiated by https://d3tglifpd8whs6.cloudfront.net/js/prebid/slash-homepage/slash-homepage.min.js.
However, in CEF, when I try to retrieve this information via the referrer URL or the Initiator URL I get "https://slashdot.org", which is not the same initiator as the browser shows me. This happens to almost all URLs, not just this one.
So, if I try to build the Initiator Chain using Referrer url or Initiator url, I am not able to get the same tree that I see in the browser.
How can I get the same initiator chain as the browser via the CEF APIs?
Or, more generically, how can I get the information to build the exact same Initiator Chain as the browser?

**Comment by amaitland** [comment]
> * I then used the DeveloperTools.Network.EnableAsync, I installed my handler for **Cefsharp.Devtools.Network.NetworkClient.RequestWillBeSent** callback, it gets invoked for every GET request and I do get the IRequest and the Initiator objects, but again they do not contain the information I need to build the Initiator chain. In many cases the Initiator URL is empty, in other cases it's just the main domain of the website.

You'll need to work out the specifics for yourself. I've never needed to recreate the initiator chain.

> Add a --devtools-protocol-log-file=<path> command-line flag for
logging protocol messages sent to/from the DevTools front-end
while it is displayed. This is useful for understanding existing
DevTools protocol usage.

`CEF` supports logging `CDP` protocol to disk, You should be able to view all commands/events to/from `CDP`

https://github.com/chromiumembedded/cef/issues/2961#issuecomment-1465023892

I'd also suggest disabling `CEF` request interceptor, it doesn't always play nicely with `CDP`.
```c#
//Disable CEF request handling, so only CDP is used.
settings.CefCommandLineArgs.Add("disable-request-handling-for-testing");
```

**Reply by thejohnbridge** [reply]
Thank you!
I added both options:

```
settings.CefCommandLineArgs.Add("--devtools-protocol-log-file=c:\\temp\\cdplog.txt");
settings.CefCommandLineArgs.Add("--disable-request-handling-for-testing");
```

but the log file is not created, I tried also removing the second option, still nothing (I'm using Windows 10 and the browser without GUI, I don't know if this makes any difference).
In general, is there any other way with CEF to track which URL is causing which GET/POST requests? I don't specifically need the initiator object, it's enough for me to keep track of which URL is causing the browser to issue which http requests.

**Reply by amaitland** [reply]
Try the following:

```c#
settings.CefCommandLineArgs.Add("devtools-protocol-log-file", "c:\\temp\\cdplog.txt");
settings.CefCommandLineArgs.Add("disable-request-handling-for-testing");
```
